### PR TITLE
Batch labels for training and inference through TED to reduce memory consumption

### PR DESCRIPTION
**Pull request by dakshvar22**
**Proposed changes**:
- Fixes https://github.com/RasaHQ/rasa/issues/8025

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [ ] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: rasa/core/policies/ted_policy.py [hunk 1]
```diff
@@ -1092,36 +1095,84 @@ def _compute_dialogue_indices(
...
+
+    def _slice_tensor(
+        self, tensor: Union[tf.Tensor, tf.SparseTensor], indices: tf.Tensor
+    ) -> Union[tf.Tensor, tf.SparseTensor]:
+        if isinstance(tensor, tf.Tensor):
+            return tf.gather_nd(tensor, tf.cast(indices, dtype=tf.int32))
+        elif isinstance(tensor, tf.SparseTensor):
+            dense_tensor = tf.sparse.to_dense(tensor)
+            sliced_tensor = tf.gather_nd(dense_tensor, tf.cast(indices, dtype=tf.int32))
+            return tf.sparse.from_dense(sliced_tensor)
```
Comment (dakshvar22) [inline rasa/core/policies/ted_policy.py:1136]: This will blow up the memory a bit but not sure what else can be done because `tf.gather_nd` does not support sparse tensors.

**Comment by Ghostvv** [comment]
could you please rebase it to `policy-train-in-chunk`, since I don't think we're going to properly finalize core chunking in the near future

File: rasa/core/policies/ted_policy.py [hunk 2]
```diff
@@ -1092,36 +1095,84 @@ def _compute_dialogue_indices(
...
+    ) -> Tuple[tf.Tensor, tf.Tensor]:
+
         all_label_ids = self.tf_label_data[LABEL_KEY][LABEL_SUB_KEY][0]
-        # labels cannot have all features "fake"
-        all_labels_encoded = {}
+
+        if number_of_random_labels != -1:
+
+            # First, sample a fixed number of random labels ids
+            total_num_label_ids = all_label_ids.shape[0]
```
Comment (Ghostvv) [inline rasa/core/policies/ted_policy.py:1107]: I think, it should be
```suggestion
            total_num_label_ids = tf.shape(all_label_ids)[0]
```
not sure if `all_label_ids` has static shape, but maybe fine if it is

File: rasa/core/policies/ted_policy.py [hunk 3]
```diff
@@ -1697,10 +1748,21 @@ def batch_loss(
...
-        label_ids = tf_batch_data[LABEL_KEY][LABEL_SUB_KEY][0]
-        labels_embed = self._get_labels_embed(label_ids, all_labels_embed)
+        (
+            sampled_label_ids,
+            sampled_labels_embed,
+        ) = self._create_all_labels_embed_for_training(
+            number_of_random_labels=self.config[LABEL_BATCH_SIZE]
+        )
+
+        batch_labels_embed = self._compute_embedding_for_label_ids(
```
Comment (Ghostvv) [inline rasa/core/policies/ted_policy.py:1760]: by doing separate computation, the number of computations are doubled compared to previous implementation

File: rasa/core/policies/ted_policy.py (hunk 2 above)
Comment (dakshvar22) [reply rasa/core/policies/ted_policy.py:1212]: `all_label_ids` has static shape, also using `tf.shape` will return a dynamic shape which cannot be used as an argument to `candidate_sampler`.
In summary, this should work as expected.

Comment (dakshvar22) [reply rasa/core/policies/ted_policy.py:1212]: `all_label_ids` has static shape, also using `tf.shape` will return a dynamic shape which cannot be used as an argument to `candidate_sampler`.
In summary, this should work as expected.

File: rasa/core/policies/ted_policy.py [hunk 4]
```diff
@@ -1179,36 +1183,179 @@ def _compute_dialogue_indices(
...
+                filtered_label_data[key] = {}
+                for sub_key in self.tf_label_data[key]:
+                    filtered_attribute_data = [
+                        self._slice_tensor(data, label_ids)
+                        for data in self.tf_label_data[key][sub_key]
+                    ]
+                    filtered_label_data[key][sub_key] = filtered_attribute_data
+
+        labels_encoded = {}
+        for key in filtered_label_data:
```
Comment (Ghostvv) [inline rasa/core/policies/ted_policy.py:1337]:
```suggestion
        for key in filtered_label_data.keys():
```

File: rasa/core/policies/ted_policy.py [hunk 5]
```diff
@@ -1179,36 +1183,179 @@ def _compute_dialogue_indices(
...
+            input_tensor.indices, sel, axis=0
+        )  # [[2,2], [2,3], [3,1]]
+        indices_new = tf.concat(
+            [
+                indices_new[:, :axis],
+                tf.expand_dims(selection_axis_indices_new, 1),
+                indices_new[:, axis + 1 :],
+            ],
+            axis=1,
+        )  # [[0,2], [0,3], [0,1]]
```
Comment (Ghostvv) [inline rasa/core/policies/ted_policy.py:1301]:
```suggestion
        )  # [[0,2], [0,3], [1,1]]
```

**Review by Ghostvv — COMMENTED** [review]
I didn't do the code review, in principle couple of variables need to be renamed and more small helper created, also the whole thing should be probably abstracted into a rasa keras layer. But idea-wise looks good to me for testing

File: rasa/utils/tensorflow/model_data.py [hunk 6]
```diff
@@ -398,12 +398,13 @@ def number_of_examples(self, data: Optional[Data] = None) -> int:
         if not example_lengths:
             return 0
 
+        # TODO: See, why this needs to be commented
```
Comment (twerkmeister) [inline rasa/utils/tensorflow/model_data.py:401]: @dakshvar22 seems this needs to be commented as the default actions have an `action_name` such as `action_listen` while the e2e bot utterances use `action_text`. If you have 1000 actions, you end up with 11 or so default actions and 989 e2e actions. And their ids get sorted into different buckets through the masks. That's why then lengths don't work out.

**Comment by twerkmeister** [comment]
Closing for now, I think this is not relevant anymore
